Fix issue where Discover tab was constantly sorting

mplorentz opened the pull request:
## Issues covered
This performance issue was blocking my work on #474 so I picked it up.

## Description
My simulator was being super slow and I found that the `FeaturedAuthorsView` was constantly sorting its authors even when the user wasn't looking at it. I think every time the viewContext was being saved swiftUI was trying to read the `authors` property, which is a computed property that does in-memory sorting and some CoreData fetch requests on the main thread.

I refactored so that instead of using an @FetchRequest for observation and then sorting the results of that in memory we keep a sorted list of IDs as @State and use `AuthorObservationView` for observation.

## How to test
1. Open the Discover tab
2. Verify that the behavior of the FeaturedAuthorsView is identical

joshuatbrown commented on Nos/Views/Discover/FeaturedAuthor.swift:
```diff
@@ -12,6 +12,21 @@ struct FeaturedAuthor {
+    
+    let rawID: RawAuthorID
+    
+    internal init(name: String, npub: String, categories: [FeaturedAuthorCategory]) {
```
```suggestion
    init(name: String, npub: String, categories: [FeaturedAuthorCategory]) {
```

`internal` isn't needed, right?

mplorentz replied:
yeah, it was auto-generated.

joshuatbrown reviewed (commented):
This is cool -- I didn't know about `AuthorObservationView`. Everything works perfectly for me. I just have a comment on the code.

joshuatbrown reviewed (changes requested):
Unfortunately, I'm having an issue on this branch where I see "Error" several times in the list on the Discover tab. This is with my account where I follow 700+ accounts, but I'm not sure if that's related.
![Screenshot 2024-06-26 at 9 41 22 AM](https://github.com/planetary-social/nos/assets/59564/5c865a8c-f7df-456f-895e-2dfb43642b85)

mplorentz commented:
@joshuatbrown oh good catch! Ok so I fixed that and absolutely blew up the diff in the process, but my changes were actually a series of simple things. Here is a summary:

- The issue you saw was that the `AuthorObservationView` doesn't create the authors if they don't exist. So I added some code to loop through and create them in `updateDisplayedFeaturedAuthors()`.
- That is great but they still show "error" for a split second, then "npub1234" for a split second, then their data loads in. This is not nice :( especially because we show a progress wheel for 2 seconds when the view loads. We should do this stuff during the progress wheel. 
- Ok I see that the progress wheel is not doing anything but wasting time 😬 it's in the DiscoverTab view and it doesn't even create `FeaturedAuthorsView` until the loading time is over. I will move this "initialLoadTime" logic into FeaturedAuthorsView and make sure that the AuthorCards are loading in the background.
- Ok now the `body` of FeaturedAuthorsView is looking really large. I will refactor the part that displays Featured Authors into a computed property... the best name seems to be `featuredAuthorsView`... ok now I'm renaming the file because it's really not only about featured authors it also does a bunch of search stuff. So now it's called `DiscoverContentView`.
- Now I see that the HomeFeedView has the same problem with the initial load time logic. I'm going to refactor it to put the `FullScreenProgressView` inside a `ZStack` so it sits on top of the ScrollView of notes. rather than using `if ... else` - just like I did in `DiscoverContentView`.
- Great now the app is deadlocking on launch. Looks like I missed a `context.perform` call in the `PushNotificationService`.

Now everything is perfect.

joshuatbrown reviewed (approved):
The code looks good, and the app works perfectly! Nice!
